### PR TITLE
[deepseek] update deepseek to real training loop, part 1

### PR DESCRIPTION
**Pull request by lessw2020**
This PR implements a core 'real' training loop in that it runs deepseekv2 model using a number of Titan components to train on real (C4) data with adamW and displays initial training loop metrics. 

There is a lot more to be done but the goal here is to get a true training loop going from which additional PRs will then improve upon it. 

Screenshot 2025-05-29 at 7 41 01 PM

A couple key highlights:
a - the model is now controllable via toml or cmd line just like Titan main.  Note that the expert parallel control is waiting for PR https://github.com/pytorch/torchtitan/pull/1244 to land...atm it just manually puts ep to 2. 
b - we use the HF deepseek tokenizer and as a result I had to make a wrapper to deal with the bos and eos params passed by Titan. 
c - loss metrics, tps, etc are displaying but MFU and tflops need to be updated. 

A lot more improvements will come shortly but for now want to land this to ensure our base deepseek training loop is available to iterate on.

### CODE REVIEW

File: torchtitan/tools/logging.py [hunk 1]
```diff
@@ -12,6 +12,10 @@
 
 
 def init_logger():
+    # Clear any existing handlers to prevent duplicate logging
```
Comment (tianyu-l) [inline torchtitan/tools/logging.py]: curious when would handlers exist before `init_logger` is called?

File: torchtitan/experiments/deepseek_v3/model_args.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+from torchtitan.components.tokenizer import Tokenizer
+from torchtitan.config_manager import JobConfig
+
+from torchtitan.protocols.train_spec import BaseModelArgs
+from torchtitan.tools.logging import logger
+from torchtitan.tools.utils import has_cuda_capability
+
+
+@dataclass
+class TransformerModelArgs(BaseModelArgs):
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/model_args.py:21]: What motivates this new class than the existing one(s) in model_config.py?

File: torchtitan/experiments/deepseek_v3/train_ds2.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    # build tokenizer
+    tokenizer = get_hf_tokenizer(model_id)
+
+    # TODO - ep is not the same as dp really...just a temp shim atm.
+    dataloader = build_hf_dataloader(
+        dp_world_size=ep_size, dp_rank=ep_rank, tokenizer=tokenizer, job_config=config
+    )
+
+    # Synthetic setting
+    microbatches = pp_size * 2
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/train_ds2.py]: Does JobConfig have a field for microbatches?

File: torchtitan/experiments/deepseek_v3/train_ds2.py
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/train_ds2.py:1]: What is the difference between train_ds.py and train_ds2.py?

File: torchtitan/experiments/deepseek_v3/__init__.py [hunk 4]
```diff
@@ -3,3 +3,53 @@
...
+        rope_theta=500000,
+    ),
+}
+
+
+register_train_spec(
+    TrainSpec(
+        name="deepseek3",
+        cls=DeepseekForCausalLM,
+        config=deepseek_configs,
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/__init__.py:46]: Maybe not a problem of this PR, but I think LHS `config` should point to a *single* model config, rather than a dictionary of model *configs*, e.g.
```
config=deepseek_debug_config,
```

File: torchtitan/experiments/deepseek_v3/infra/parallelize_deepseek.py
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/infra/parallelize_deepseek.py:1]: Perhaps no need to put it under an `infra` folder? I don't see how they are related.

File: torchtitan/experiments/deepseek_v3/train_ds2.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+from torchtitan.experiments.deepseek_v3.tokenizers.hf_tokenizer import get_hf_tokenizer
+
+# from torchtitan.experiments.deepseek_v3.train_configs.custom_args import JobConfig
+from torchtitan.tools import utils
+from torchtitan.tools.logging import init_logger, logger
+from torchtitan.tools.utils import get_device_info
+
+# Use DeepSeek-V2-Lite as a proxy
+model_id = "deepseek-ai/DeepSeek-V2-Lite"
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/train_ds2.py]: maybe `model_id` should come from `JobConfig`?

File: torchtitan/experiments/deepseek_v3/train_ds2.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    # metrics manager
+    proxy_parallel_dims = ParallelDims(
+        dp_replicate=ep_size,
+        dp_shard=fsdp_dim,
+        pp=pp_size,
+        cp=1,
+        tp=1,
+        world_size=world_mesh.size(),
+        enable_loss_parallel=False,
+    )
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/train_ds2.py]: This looks like a duplicate of information of what `DeviceMesh` or `config.parallelism` would carry.

File: torchtitan/experiments/deepseek_v3/train_ds2.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    # TODO - remove this for full model
+    # model_args.num_hidden_layers = 16
+
+    (
+        model,
+        pp_size,
+        pp_rank,
+        pp_mesh,
+        ep_size,
+        ep_rank,
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/train_ds2.py]: logically, aren't these pre-known, or known easier via:
```
device_mesh.get_rank(dim="pp")
```
?

File: torchtitan/experiments/deepseek_v3/__init__.py [hunk 8]
```diff
@@ -3,3 +3,53 @@
...
+        parallelize_fn=parallelize_deepseek,
+        pipelining_fn=pipeline_llama,
+        build_optimizers_fn=build_optimizers,
+        build_lr_schedulers_fn=build_lr_schedulers,
+        build_dataloader_fn=build_hf_dataloader,
+        build_tokenizer_fn=get_hf_tokenizer,
+        build_loss_fn=build_cross_entropy_loss,
```
Comment (kwen2501) [inline torchtitan/experiments/deepseek_v3/__init__.py:53]: It seems to make more sense to directly invoke these "build_..." functions in train.py. `JobConfig` should be more for int, str, float, etc.
If changed, the imports of these functions at top can be moved away too. (would make `__init__.py` much cleaner imo)

File: torchtitan/tools/logging.py (hunk 1 above)
Comment (lessw2020) [reply torchtitan/tools/logging.py]: for this PR it does, but I have not been able to track down why or where the other handler comes from. 
It also does the same thing in my earlier WIP PR...one theory is HF tokenizer is adding a base handler, but I haven't pinned it down and it's non obvious (i.e. I did go through looking for other logging instances). 

What I do know is that without this clearance, the user will get everything repeated x2 - one from base handler and then one from [titan].
I can try to isolate this better but at the same time generically if we want to ensure only one logging is in place (titan's) then this also accomplishes a defensive setup.

File: torchtitan/experiments/deepseek_v3/train_ds2.py
Comment (lessw2020) [reply torchtitan/experiments/deepseek_v3/train_ds2.py:1]: train_ds is the original synthetic data, no optimizer etc. prototype training loop.  
ds2 is the 'real' data training loop with real data, dataloader, optimizer, config, etc.
I wasn't clear if we wanted to leave the prototype training loop available for dev or debugging given it's very simple vs. real training loop adds a lot of features.

File: torchtitan/experiments/deepseek_v3/infra/parallelize_deepseek.py
Comment (lessw2020) [reply torchtitan/experiments/deepseek_v3/infra/parallelize_deepseek.py:1]: I'll move it - I was mostly just trying to keep some similarity with the llama4 layout.

File: torchtitan/experiments/deepseek_v3/train_ds2.py
Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/train_ds2.py:1]: Makes sense! Maybe reflect the above in the file names while we don't have a decision?

Comment (lessw2020) [reply torchtitan/experiments/deepseek_v3/train_ds2.py:1]: great idea - renamed to train_ds_dev and train_ds_real for easy id of what is what.
Also for reference, I changed from generic train to train_ds_[dev/real] b/c it gets confusing having so many 'train.py" running around in the project (main, llama4, etc). 
Anyway, updated to much more intuitive file naming, thanks for the feedback/idea.

File: torchtitan/tools/logging.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/tools/logging.py]: I see. I suggest we don't include changes which we don't fully understand yet.

We can open a separate PR / issue fixing the problem when we have enough understanding.

To unblock, you could do this cleaning inside the experiment folder where HF tokenizer is used.

Comment (lessw2020) [reply torchtitan/tools/logging.py]: I've confirmed the additional handler is from HF tokenizer so it is now fully isolated.  
Removed this change to clear out other handlers from titan logging, will see if I can clear it in experimental.

**Comment by lessw2020** [comment]
GPU CI failure is not related (hit exact same error on earlier PR, which also was unrelated).
